src/core/model/table/column_domain.cpp

```diff
@@ -0,0 +1,306 @@
+/** \file
+ * \brief Column domain data
+ *
+ * implementation of partition reader classes and domain creation manager
+ */
+#include "column_domain.h"
+
+#include <algorithm>
+#include <cmath>
+#include <numeric>
+#include <string>
+
+#include <easylogging++.h>
+
+#include "config/thread_number/type.h"
+#include "model/table/block_dataset_stream.h"
+#include "util/parallel_for.h"
+
+namespace model {
+
+using PartitionReader = DomainPartition::PartitionReader;
+
+/// reader for reading data from main memory
+class MemoryBackedReader final : public PartitionReader {
+public:
+    using Values = DomainPartition::Values;
+
+private:
+    using It = Values::iterator;
+
+    It cur_, end_;
+
+public:
+    explicit MemoryBackedReader(Values const& values) : cur_(values.begin()), end_(values.end()) {
+        assert(cur_ != end_);
+    }
+
+    Value const& GetValue() const final {
+        return *cur_;
+    }
+
+    bool HasNext() const final {
+        return std::next(cur_) != end_;
+    }
+
+    void MoveToNext() final {
+        ++cur_;
+    }
+};
+
+/// reader for reading data from swap file
+class FileBackedReader final : public PartitionReader {
+private:
+    std::ifstream file_;
+    Value cur_;
+
+public:
+    explicit FileBackedReader(std::filesystem::path const& path) : file_(path) {
+        if (!file_.is_open()) {
+            throw std::runtime_error("Error opening file");
+        }
+        MoveToNext();
+    }
+
+    Value const& GetValue() const final {
+        return cur_;
+    }
+
+    bool HasNext() const final {
+        return !file_.eof();
+    }
+
+    void MoveToNext() final {
+        std::getline(file_, cur_);
+    }
+};
+
+bool PartitionReader::TryMove() {
+    bool can_move = HasNext();
+    if (can_move) MoveToNext();
+    return can_move;
+}
+
+DomainPartition::~DomainPartition() {
+    if (IsSwapped()) {
+        std::filesystem::remove(*swap_file_);
+    }
+}
+
+std::unique_ptr<PartitionReader> DomainPartition::GetReader() const {
+    if (IsSwapped()) {
+        return std::make_unique<FileBackedReader>(*swap_file_);
+    } else {
+        return std::make_unique<MemoryBackedReader>(values_);
+    }
+}
+
+size_t DomainPartition::GetMemoryUsage() const {
+    /* in reality, partition storage takes on average
+     * 5 times more memory than the presented estimate
+     */
+    static constexpr double kAverageOverhead = 5.0;
+    size_t total = 0;
```

Comment:
```suggestion
    if (IsSwapped()) return 0;
    const size_t data_size = std::accumulate(values_ | std::views::transform([](const std::string& str){ return str.capacity(); }), 0);
    return (data_size + values_.size() * sizeof(Values::node_type)) * kAverageOverhead;
```

Reply:
Estimating memory consumption is the most difficult part. In fact, we rely on a specific implementation, which we would like to avoid. Ideally, it seems to me that we should monitor memory usage in another way (maybe use `std::pmr::set` with custom allocator).
I also made some changes to the code that made the estimation more accurate.
Regarding `str.size() >= str.capacity()`, this is an error. There should be a `!=` check here. We assume that the compiler has SSO (small string optimization). And enough small strings should fit completely into the `std::string`. If the SSO works for string, then the size will not match the capacity. Otherwise, we take the capacity, since in this case memory allocation occurred on the heap.
So, `data_total_size` will contain memory usage in bytes for string's data, which was allocated in the heap.
And it also remains to take into account how much memory is spent on the nodes themselves. Here we multiply the number of nodes by the sum of the `sizeof(std::string)` and `sizeof(std::set<std::string>::node_type)`. And we also have to multiply this estimate by the coefficient that was derived experimentally.

I will also give an example of how this estimation works on TPCH dataset (with scale factor 1):
```
memory usage (estimate): 0 mb, current rss: 8.625 mb
memory usage (estimate): 93 mb, current rss: 101.098 mb
memory usage (estimate): 309 mb, current rss: 317.098 mb
memory usage (estimate): 827 mb, current rss: 846.598 mb
memory usage (estimate): 827 mb, current rss: 846.598 mb
memory usage (estimate): 1214 mb, current rss: 1224.1 mb
memory usage (estimate): 1277 mb, current rss: 1285.97 mb
memory usage (estimate): 1331 mb, current rss: 1339.6 mb
memory usage (estimate): 1520 mb, current rss: 1533.8 mb
memory usage (estimate): 1520 mb, current rss: 1533.8 mb
```
Left is estimation from `GetMemoryUsage()`, right is current RSS.
In this case we get a fairly good estimate.